### ISSUE TITLE
Inference result

### ISSUE THREAD

**Issue by AJITKUMAR130012** [ultralytics/ultralytics#19225]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

   while True:
        ret, frame = cap.read()
        if not ret:
            break
        #frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)

        frame_resized = cv2.resize(frame, (640, 640))
        
        #gpu_tensor = torch.tensor(frame_resized, dtype=torch.float32).permute(2, 0, 1).unsqueeze(0)  # Shape: (1,3,640,640)
        gpu_tensor = torch.from_numpy(frame_resized)
        gpu_tensor= gpu_tensor.float()

        gpu_tensor /= 255.0
        gpu_tensor = gpu_tensor.permute(2, 0, 1).unsqueeze(0)
        # Convert to float32 tensor in correct format
        #
        #gpu_tensor = gpu_tensor.to("cuda")

        # Pass both the GPU tensor and original frame
        results = model.track(gpu_tensor, persist=True)

When i tried to inference the by converting the frame(numpy) to the pytorch tensor to utilize the GPU in JetsonAGX Xavier, then it gives no detection
Result after converting to the tensor:

![Image](https://github.com/user-attachments/assets/0705358e-f5a6-4d1a-a109-ebcd3d4d7a28)

Result for the frame(numpy-array):

![Image](https://github.com/user-attachments/assets/29bb66b8-aa74-46e5-8fae-2023c4e1a319)

So, what correction do i need to utilize the GPU while inferenece to reduce the inference time?

### Environment

neuron@ubuntu:~/Documents/Ajit/pysparks$ jetson_release
Software part of jetson-stats 4.2.8 - (c) 2024, Raffaello Bonghi
Model: NVIDIA Jetson Xavier NX Developer Kit - Jetpack 5.1.1 [L4T 35.3.1]
NV Power Mode[0]: MODE_15W_2CORE
Serial Number: [XXX Show with: jetson_release -s XXX]
Hardware:
 - P-Number: p3668-0001
 - Module: NVIDIA Jetson Xavier NX
Platform:
 - Distribution: Ubuntu 20.04 focal
 - Release: 5.10.104-tegra
jtop:
 - Version: 4.2.8
 - Service: Active
Libraries:
 - CUDA: 11.4.315
 - cuDNN: 8.6.0.166
 - TensorRT: 8.5.2.2
 - VPI: 2.2.7
 - OpenCV: 4.11.0 - with CUDA: YES
neuron@ubuntu:~/Documents/Ajit/pysparks$ uname -a
Linux ubuntu 5.10.104-tegra #1 SMP PREEMPT Sun Mar 19 07:55:28 PDT 2023 aarch64 aarch64 aarch64 GNU/Linux

### Minimal Reproducible Example

while True:
        ret, frame = cap.read()
        if not ret:
            break
        #frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)

        frame_resized = cv2.resize(frame, (640, 640))
        
        #gpu_tensor = torch.tensor(frame_resized, dtype=torch.float32).permute(2, 0, 1).unsqueeze(0)  # Shape: (1,3,640,640)
        gpu_tensor = torch.from_numpy(frame_resized)
        gpu_tensor= gpu_tensor.float()

        gpu_tensor /= 255.0
        gpu_tensor = gpu_tensor.permute(2, 0, 1).unsqueeze(0)
        # Convert to float32 tensor in correct format
        #
        #gpu_tensor = gpu_tensor.to("cuda")

        # Pass both the GPU tensor and original frame
        results = model.track(gpu_tensor, persist=True)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
👋 Hello @AJITKUMAR130012, thank you for sharing your detailed issue regarding inference on the Jetson Xavier NX 🚀! We truly appreciate your thorough report, including the code snippet, system environment details, and output images—it’s incredibly helpful for us to investigate.

### Quick Notes
It seems you're encountering an issue when converting a `numpy` array to a PyTorch tensor for GPU inference, resulting in no detections. This might be related to tensor preparation or device placement. Let’s address this further with some follow-up actions.

### Bug 🐛
If this is a 🐛 Bug Report, kindly provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us reliably debug the issue. While you've shared the main code block, it'd be great if you could confirm additional context such as:
- How the `model` object is initialized (e.g., is it loaded with `.load()` for tracking specifically?).
- Any modifications made to the `track` method, if applicable.

Additionally, ensure your `ultralytics` package is up to date. Run the code below to upgrade to the latest version:

```bash
pip install -U ultralytics
```

### GPU Environment
Since you are operating on Jetpack with CUDA and cuDNN, ensure the correct configuration of PyTorch, as Jetson devices often require a custom PyTorch installation. Verify by running the following:

```python
import torch
print(torch.cuda.is_available())  # Should return True for GPU availability
print(torch.__version__)  # Validate compliance with your CUDA version
```

You may also replace `.to("cuda")` with `.to(torch.device("cuda"))` to ensure compatibility on your specific hardware.

### Resources 🛠️
Here are some helpful docs to troubleshoot and refine your approach:
- [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum-reproducible-example/)
- [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)
- [PyTorch and CUDA Setup](https://pytorch.org/get-started/locally/)

### Join the Community 💬
Let’s keep the conversation going! Engage with the Ultralytics community for support and discussions:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com/) for forums
- [Subreddit](https://reddit.com/r/Ultralytics) to share insights with fellow users

### Additional Debugging
For more seamless debugging of live inference:
1. Try running the model on one of our verified environments:
    - <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>
    - <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
    - <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
2. Confirm the robustness of your setup with our verified YOLO usage guides in the [Docs](https://docs.ultralytics.com/).

---

This is an automated response, but don’t worry—an Ultralytics engineer will review and assist you further as soon as possible. Thank you for your patience and for being part of the Ultralytics community 😊!

**Comment by AJITKUMAR130012** [comment]
Inference time is : 18.13996911048889s when i moved to the torch tensor

![Image](https://github.com/user-attachments/assets/2afd6e7f-9db5-4f8d-8c27-5d6347dd86b4)

Inference time is : 28.256394386291504 when i directly inference on the frame(numpy-array)

![Image](https://github.com/user-attachments/assets/b09d2d82-a99e-42ca-ae9b-6007864dce06)

So, this makes the inference faster.


> The preprocessing is wrong. And I don't see the point here. Ultralytics moves it to the GPU internally. You're not making it any faster. Especially since the resizing is not being performed on GPU.

**Comment by AJITKUMAR130012** [comment]
> 👋 Hello [@AJITKUMAR130012](https://github.com/AJITKUMAR130012), thank you for sharing your detailed issue regarding inference on the Jetson Xavier NX 🚀! We truly appreciate your thorough report, including the code snippet, system environment details, and output images—it’s incredibly helpful for us to investigate.
> 
> ### Quick Notes
> It seems you're encountering an issue when converting a `numpy` array to a PyTorch tensor for GPU inference, resulting in no detections. This might be related to tensor preparation or device placement. Let’s address this further with some follow-up actions.
> 
> ### Bug 🐛
> If this is a 🐛 Bug Report, kindly provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us reliably debug the issue. While you've shared the main code block, it'd be great if you could confirm additional context such as:
> 
> * How the `model` object is initialized (e.g., is it loaded with `.load()` for tracking specifically?).
> * Any modifications made to the `track` method, if applicable.
> 
> Additionally, ensure your `ultralytics` package is up to date. Run the code below to upgrade to the latest version:
> 
> pip install -U ultralytics
> ### GPU Environment
> Since you are operating on Jetpack with CUDA and cuDNN, ensure the correct configuration of PyTorch, as Jetson devices often require a custom PyTorch installation. Verify by running the following:
> 
> import torch
> print(torch.cuda.is_available())  # Should return True for GPU availability
> print(torch.__version__)  # Validate compliance with your CUDA version
> You may also replace `.to("cuda")` with `.to(torch.device("cuda"))` to ensure compatibility on your specific hardware.
> 
> ### Resources 🛠️
> Here are some helpful docs to troubleshoot and refine your approach:
> 
> * [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum-reproducible-example/)
> * [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)
> * [PyTorch and CUDA Setup](https://pytorch.org/get-started/locally/)
> 
> ### Join the Community 💬
> Let’s keep the conversation going! Engage with the Ultralytics community for support and discussions:
> 
> * [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
> * [Discourse](https://community.ultralytics.com/) for forums
> * [Subreddit](https://reddit.com/r/Ultralytics) to share insights with fellow users
> 
> ### Additional Debugging
> For more seamless debugging of live inference:
> 
> 1. Try running the model on one of our verified environments:
>    
>    * [![Run on Gradient](https://camo.githubusercontent.com/6df71b7d7e0b09a2e97776f416bcd40acf48fc87337f8bcce48e4235537daf1f/68747470733a2f2f6173736574732e706170657273706163652e696f2f696d672f6772616469656e742d62616467652e737667)](https://console.paperspace.com/github/ultralytics/ultralytics)
>    * [![Open In Colab](https://camo.githubusercontent.com/96889048f8a9014fdeba2a891f97150c6aac6e723f5190236b10215a97ed41f3/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)
>    * [![Open In Kaggle](https://camo.githubusercontent.com/c7135949c5c6882489e68f4af05a78a759460a4db256b86df3097e04419b4d9e/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolo11)
> 2. Confirm the robustness of your setup with our verified YOLO usage guides in the [Docs](https://docs.ultralytics.com/).
> 
> This is an automated response, but don’t worry—an Ultralytics engineer will review and assist you further as soon as possible. Thank you for your patience and for being part of the Ultralytics community 😊!


This is the code snipts for load the model that is present in local directory:


# model=YOLO('traffic_best_n.pt')
model = YOLO("/home/neuron/Documents/Ajit/pysparks/traffic_best_n.engine")
my_file = open("/home/neuron/Documents/Ajit/pysparks/labels.txt", "r")
data = my_file.read()
class_list = data.split("\n")


No any modification, i have done in the track methode.


This is screenshot for the GPU availability:

![Image](https://github.com/user-attachments/assets/ee08af84-4c5f-409c-9ffd-86e12e6afd12)

**Comment by Y-T-G** [comment]
> So, this makes the inference faster.

This is because you're repeatedly inferring on the same tensor with preprocessing performed once. But that's not how real-life inference works. Each frame is different, and each frame will need to be preprocessed separately. Your calculated time is not representative.

Your approach wouldn't result in any improvement in actual inference.

**Comment by Y-T-G** [comment]
> So, this makes the inference faster.

This is because the preprocessing is wrong. Ultralytics uses Letterbox resizing. You're using naive resizing. And it's also because nothing is detected with your preprocessed tensor inference, so there's no post-processing to be done which makes it faster.

**Comment by AJITKUMAR130012** [comment]
When i am send the gpu tensor without resizing then i got the following code and error.

![Image](https://github.com/user-attachments/assets/e52b61e2-916b-4f61-820c-1f19dc74b0a6)

![Image](https://github.com/user-attachments/assets/7ece6527-78f9-46c0-99c8-82906b3005ee)

**Comment by Y-T-G** [comment]
There's no point in putting it on GPU. Ultralytics does automatically that like I said after resizing and preprocessing.

**Comment by Y-T-G** [comment]
The correct pre-processing is shown here except you also need to add `permute()`.

https://github.com/ultralytics/ultralytics/blob/d317b26940e4c064bd862cd5cfe426aaf76c41be/examples/YOLOv8-TFLite-Python/main.py#L126

**Comment by Y-T-G** [comment]
If you want faster inference, you can try this instead.

```python
from ultralytics import YOLO

# Load your own model. This is an example.
model = YOLO("yolov8n.pt")

# Run it only once outside the loop to create the predictor.
model.track(frame, save=False, show=False, conf=0.01, persist=True) # Add other parameters that you use
predictor = model.predictor

# Use this instead of model.track in loop
prep = predictor.preprocess([frame])
output = predictor.inference(prep)
predictor.results = predictor.postprocess(output, prep, [frame])
predictor.run_callbacks("on_predict_postprocess_end")
results = predictor.results
```

**Comment by Y-T-G** [comment]
And if you want faster tracking use `tracker="bytetrack.yaml"`.

Or use this Rust version.

https://github.com/Y-T-G/ioutrack_rs
